### ISSUE TITLE
Update for ember-cli 2.10

### ISSUE THREAD

**Issue by xcambar** [201-created/ember-cli-acceptance-test-helpers#34]
While upgrading to ember-cli 2.10, I noticed the addon is incompatible as it relies on globally available `ok` (see https://github.com/201-created/ember-cli-acceptance-test-helpers/blob/master/test-support/helpers/201-created/utils/wrap-in-expectation.js).

I could not find how to get access to assertions from the test helper registration functions, and moreover, the Ember docs state that `assert` and other must be passed explicitly to the helper (see https://guides.emberjs.com/v2.10.0/testing/acceptance/#toc_custom-test-helpers).

As far as I understand and in order to have this addon work with the latest ember-cli, a change in API seems necessary, either as:
* rename `expectComponent` (and others) to `findComponents` and use it as `assert.ok(findComponents("my-component"))` or `assert.equal(findComponents("my-component").length, 3)`
* prepend `assert` as the first parameter to the existing components `expectComponents(assert, "my-component")`

Obviously, more options are available, pros and cons will be discussed, please bare in mind the two above are just a way to kickstart the discussion.

What do you think?

**Comment by xcambar** [comment]
Sorry, I was too quick to write this issue, I missed https://github.com/201-created/ember-cli-acceptance-test-helpers/pull/33

I'll leave open as the PR is not merged yet.

**Comment by jelhan** [comment]
I like the `assert.ok(findComponents("my-component"))` and `assert.equals(findComponents("my-component").length, 3)` style. Not quite sure what findComponents should return than. A [NodeListe](https://developer.mozilla.org/en-US/docs/Web/API/NodeList)? A jQuery collection? This might allow us to use same API in acceptance and component integration tests.

**Comment by jelhan** [comment]
@iezer @bantic: Which syntax would you prefer? I'm willing to rewrite #33 accordingly to your decision if you are willing to merge. Maybe we could support both. `findComponent(componentName, options)` to return a jQuery collection of component nodes or NodeList and `expectComponent(assert, componentName, count, options)`. If you insist in backward compatibility we could also support current `expectCompoent(componentName, count, options)` syntax for ember-cli < 2.10 by parsing arguments. Just let me know what your thoughts are.

**Comment by iezer** [comment]
@jelhan and @xcambar What do you think about adding the "expect" helpers onto the assert prototype, so you could do `assert.expectElement()` and `assert.expectComponent()`?

**Comment by xcambar** [comment]
Seems a bit off to have `assert.expect...`, IMO.

Can you elaborate on the rationale behind the idea?

**Comment by iezer** [comment]
Adding custom assertions is a common technique and [documented in QUnit](https://qunitjs.com/cookbook/#custom-assertions). I like this syntax because it's similar to existing QUnit helpers so it makes our helpers here feel less like a special case and more like a part of the test suite. 

I agree that `assert.expect` redundant and confusing. We can rename `expect` to something like `has`. So our API would be `assert.hasComponent`, `assert.hasElement`, `assert.hasNoComponent`, `assert.hasNoElement`.

For example:

```js
assert.equal(val, expect, 'message');
assert.hasComponent('power-select'); // consistent usage
```

```js
assert.equal(val, expected, 'message');
expectComponent(assert, 'power-select')); // what's with this special case for expectComponent passing in assert?

```

The only problem is that this syntax wouldn't work for mocha. I think we can punt on that because I'm not sure many people use this library with mocha and there is already a fork.

Also sometimes people call `expectElement` with the result of `find` for example `expectElement(find('Link To Home'));` so we would have to make sure to support `assert.hasElement(find('Link To Home'))`.

**Comment by xcambar** [comment]
I like the `has` syntax much better, FWIW!
Overall, I agree with all your points. Thanks!

**Comment by jelhan** [comment]
`assert.hasComponent(componentName, count, options)` seems nice. But do we really have access to current application in a custom assertion?

How would the syntax for integration tests look like? I guess we still would have to pass container as first argument? So it would be `assert.hasComponent(this.container, componentName, count, options)`?

**Comment by jelhan** [comment]
It should take `message` as an optional last parameter. So full syntax for acceptance tests should be:

```
assert.hasComponent(componentName, count, options, message);
assert.hasComponent(componentName, count, options);
assert.hasComponent(componentName, count, message);
assert.hasComponent(componentName, count);
assert.hasComponent(componentName, message);
assert.hasComponent(componentName);
```

I'm not quite sure if we should also support this two:

```
assert.hasComponent(componentName, options, message);
 assert.hasComponent(componentName, options);
```

I guess parsing the arguments would be easy. `componentName` must be the first argument. The other arguments could be determined by type: a `number` is `count`, a `string` is `message` and an `object` is `options` hash.

**Comment by jelhan** [comment]
@iezer Thanks for your reply at https://github.com/201-created/ember-cli-acceptance-test-helpers/pull/33#issuecomment-290752322.

If we decide to implement `hasComponent` and `hasElement` as custom qunit assertions we should rely on [ember-cli-custom-assertions](https://github.com/DockYard/ember-cli-custom-assertions). I didn't used it yet but it looks great. Especially it provides context as first argument in custom assertions which is what we need. But it's not stable yet (released 0.0.7 a half year ago). Would you accept such a rewrite?

**Comment by iezer** [comment]
Fixed in version 1.0!
